{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "This event loop is already running",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 47\u001b[0m\n\u001b[0;32m     44\u001b[0m                 \u001b[39mbreak\u001b[39;00m\n\u001b[0;32m     46\u001b[0m \u001b[39m# Run the async function\u001b[39;00m\n\u001b[1;32m---> 47\u001b[0m asyncio\u001b[39m.\u001b[39;49mget_event_loop()\u001b[39m.\u001b[39;49mrun_until_complete(generate_text())\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\asyncio\\base_events.py:625\u001b[0m, in \u001b[0;36mBaseEventLoop.run_until_complete\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    614\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Run until the Future is done.\u001b[39;00m\n\u001b[0;32m    615\u001b[0m \n\u001b[0;32m    616\u001b[0m \u001b[39mIf the argument is a coroutine, it is wrapped in a Task.\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    622\u001b[0m \u001b[39mReturn the Future's result, or raise its exception.\u001b[39;00m\n\u001b[0;32m    623\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    624\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_closed()\n\u001b[1;32m--> 625\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_check_running()\n\u001b[0;32m    627\u001b[0m new_task \u001b[39m=\u001b[39m \u001b[39mnot\u001b[39;00m futures\u001b[39m.\u001b[39misfuture(future)\n\u001b[0;32m    628\u001b[0m future \u001b[39m=\u001b[39m tasks\u001b[39m.\u001b[39mensure_future(future, loop\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\asyncio\\base_events.py:584\u001b[0m, in \u001b[0;36mBaseEventLoop._check_running\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    582\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_check_running\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m    583\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mis_running():\n\u001b[1;32m--> 584\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\u001b[39m'\u001b[39m\u001b[39mThis event loop is already running\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m    585\u001b[0m     \u001b[39mif\u001b[39;00m events\u001b[39m.\u001b[39m_get_running_loop() \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    586\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\n\u001b[0;32m    587\u001b[0m             \u001b[39m'\u001b[39m\u001b[39mCannot run the event loop while another loop is running\u001b[39m\u001b[39m'\u001b[39m)\n",
      "\u001b[1;31mRuntimeError\u001b[0m: This event loop is already running"
     ]
    }
   ],
   "source": [
    "import asyncio\n",
    "import json\n",
    "import websockets\n",
    "\n",
    "async def generate_text():\n",
    "    async with websockets.connect('wss://chat.petals.dev/api/v2/generate') as websocket:\n",
    "        # Send open_inference_session request\n",
    "        open_session_request = {\n",
    "            \"type\": \"open_inference_session\",\n",
    "            \"max_length\": 1024\n",
    "        }\n",
    "        await websocket.send(json.dumps(open_session_request))\n",
    "        \n",
    "        response = await websocket.recv()\n",
    "        response_data = json.loads(response)\n",
    "        \n",
    "        if response_data.get(\"ok\"):\n",
    "            print(\"Inference session opened successfully.\")\n",
    "        else:\n",
    "            print(\"Failed to open inference session:\", response_data.get(\"traceback\"))\n",
    "            return\n",
    "        \n",
    "        # Send generate request\n",
    "        generate_request = {\n",
    "            \"type\": \"generate\",\n",
    "            \"inputs\": \"A cat in French is \\\"\",\n",
    "            \"max_new_tokens\": 3\n",
    "        }\n",
    "        await websocket.send(json.dumps(generate_request))\n",
    "        \n",
    "        while True:\n",
    "            response = await websocket.recv()\n",
    "            response_data = json.loads(response)\n",
    "            \n",
    "            if response_data.get(\"ok\"):\n",
    "                outputs = response_data.get(\"outputs\")\n",
    "                stop = response_data.get(\"stop\")\n",
    "                print(\"Generated:\", outputs)\n",
    "                if stop:\n",
    "                    print(\"Generation completed.\")\n",
    "                    break\n",
    "            else:\n",
    "                print(\"Failed to generate text:\", response_data.get(\"traceback\"))\n",
    "                break\n",
    "\n",
    "# Run the async function\n",
    "asyncio.get_event_loop().run_until_complete(generate_text())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "TimeoutError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mCancelledError\u001b[0m                            Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\site-packages\\websockets\\legacy\\client.py:655\u001b[0m, in \u001b[0;36mConnect.__await_impl_timeout__\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    654\u001b[0m \u001b[39masync\u001b[39;00m \u001b[39mwith\u001b[39;00m asyncio_timeout(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mopen_timeout):\n\u001b[1;32m--> 655\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m__await_impl__()\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\site-packages\\websockets\\legacy\\client.py:659\u001b[0m, in \u001b[0;36mConnect.__await_impl__\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    658\u001b[0m \u001b[39mfor\u001b[39;00m redirects \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mMAX_REDIRECTS_ALLOWED):\n\u001b[1;32m--> 659\u001b[0m     _transport, _protocol \u001b[39m=\u001b[39m \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_create_connection()\n\u001b[0;32m    660\u001b[0m     protocol \u001b[39m=\u001b[39m cast(WebSocketClientProtocol, _protocol)\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\asyncio\\base_events.py:1060\u001b[0m, in \u001b[0;36mBaseEventLoop.create_connection\u001b[1;34m(self, protocol_factory, host, port, ssl, family, proto, flags, sock, local_addr, server_hostname, ssl_handshake_timeout, happy_eyeballs_delay, interleave)\u001b[0m\n\u001b[0;32m   1059\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m-> 1060\u001b[0m     sock \u001b[39m=\u001b[39m \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_connect_sock(\n\u001b[0;32m   1061\u001b[0m         exceptions, addrinfo, laddr_infos)\n\u001b[0;32m   1062\u001b[0m     \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\asyncio\\base_events.py:969\u001b[0m, in \u001b[0;36mBaseEventLoop._connect_sock\u001b[1;34m(self, exceptions, addr_info, local_addr_infos)\u001b[0m\n\u001b[0;32m    968\u001b[0m             \u001b[39mraise\u001b[39;00m \u001b[39mOSError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mno matching local address with \u001b[39m\u001b[39m{\u001b[39;00mfamily\u001b[39m=}\u001b[39;00m\u001b[39m found\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m--> 969\u001b[0m \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msock_connect(sock, address)\n\u001b[0;32m    970\u001b[0m \u001b[39mreturn\u001b[39;00m sock\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\asyncio\\selector_events.py:501\u001b[0m, in \u001b[0;36mBaseSelectorEventLoop.sock_connect\u001b[1;34m(self, sock, address)\u001b[0m\n\u001b[0;32m    500\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 501\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mawait\u001b[39;00m fut\n\u001b[0;32m    502\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    503\u001b[0m     \u001b[39m# Needed to break cycles when an exception occurs.\u001b[39;00m\n",
      "\u001b[1;31mCancelledError\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mTimeoutError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 47\u001b[0m\n\u001b[0;32m     44\u001b[0m                 \u001b[39mbreak\u001b[39;00m\n\u001b[0;32m     46\u001b[0m \u001b[39m# Run the async function using await\u001b[39;00m\n\u001b[1;32m---> 47\u001b[0m \u001b[39mawait\u001b[39;00m generate_text()\n",
      "Cell \u001b[1;32mIn[3], line 6\u001b[0m, in \u001b[0;36mgenerate_text\u001b[1;34m()\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[39masync\u001b[39;00m \u001b[39mdef\u001b[39;00m \u001b[39mgenerate_text\u001b[39m():\n\u001b[1;32m----> 6\u001b[0m     \u001b[39masync\u001b[39;00m \u001b[39mwith\u001b[39;00m websockets\u001b[39m.\u001b[39mconnect(\u001b[39m'\u001b[39m\u001b[39mwss://chat.petals.dev/api/v2/generate\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mas\u001b[39;00m websocket:\n\u001b[0;32m      7\u001b[0m         \u001b[39m# Send open_inference_session request\u001b[39;00m\n\u001b[0;32m      8\u001b[0m         open_session_request \u001b[39m=\u001b[39m {\n\u001b[0;32m      9\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mtype\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39m\"\u001b[39m\u001b[39mopen_inference_session\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m     10\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mmax_length\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39m1024\u001b[39m\n\u001b[0;32m     11\u001b[0m         }\n\u001b[0;32m     12\u001b[0m         \u001b[39mawait\u001b[39;00m websocket\u001b[39m.\u001b[39msend(json\u001b[39m.\u001b[39mdumps(open_session_request))\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\site-packages\\websockets\\legacy\\client.py:637\u001b[0m, in \u001b[0;36mConnect.__aenter__\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    636\u001b[0m \u001b[39masync\u001b[39;00m \u001b[39mdef\u001b[39;00m \u001b[39m__aenter__\u001b[39m(\u001b[39mself\u001b[39m) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m WebSocketClientProtocol:\n\u001b[1;32m--> 637\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\site-packages\\websockets\\legacy\\client.py:654\u001b[0m, in \u001b[0;36mConnect.__await_impl_timeout__\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    653\u001b[0m \u001b[39masync\u001b[39;00m \u001b[39mdef\u001b[39;00m \u001b[39m__await_impl_timeout__\u001b[39m(\u001b[39mself\u001b[39m) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m WebSocketClientProtocol:\n\u001b[1;32m--> 654\u001b[0m     \u001b[39masync\u001b[39;00m \u001b[39mwith\u001b[39;00m asyncio_timeout(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mopen_timeout):\n\u001b[0;32m    655\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m__await_impl__()\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\site-packages\\websockets\\legacy\\async_timeout.py:169\u001b[0m, in \u001b[0;36mTimeout.__aexit__\u001b[1;34m(self, exc_type, exc_val, exc_tb)\u001b[0m\n\u001b[0;32m    163\u001b[0m \u001b[39masync\u001b[39;00m \u001b[39mdef\u001b[39;00m \u001b[39m__aexit__\u001b[39m(\n\u001b[0;32m    164\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m    165\u001b[0m     exc_type: Optional[Type[\u001b[39mBaseException\u001b[39;00m]],\n\u001b[0;32m    166\u001b[0m     exc_val: Optional[\u001b[39mBaseException\u001b[39;00m],\n\u001b[0;32m    167\u001b[0m     exc_tb: Optional[TracebackType],\n\u001b[0;32m    168\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Optional[\u001b[39mbool\u001b[39m]:\n\u001b[1;32m--> 169\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_do_exit(exc_type)\n\u001b[0;32m    170\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\anton_glfk00f\\miniconda3\\envs\\llm_app_env\\lib\\site-packages\\websockets\\legacy\\async_timeout.py:252\u001b[0m, in \u001b[0;36mTimeout._do_exit\u001b[1;34m(self, exc_type)\u001b[0m\n\u001b[0;32m    250\u001b[0m \u001b[39mif\u001b[39;00m exc_type \u001b[39mis\u001b[39;00m asyncio\u001b[39m.\u001b[39mCancelledError \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39m==\u001b[39m _State\u001b[39m.\u001b[39mTIMEOUT:\n\u001b[0;32m    251\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_timeout_handler \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m--> 252\u001b[0m     \u001b[39mraise\u001b[39;00m asyncio\u001b[39m.\u001b[39mTimeoutError\n\u001b[0;32m    253\u001b[0m \u001b[39m# timeout has not expired\u001b[39;00m\n\u001b[0;32m    254\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39m=\u001b[39m _State\u001b[39m.\u001b[39mEXIT\n",
      "\u001b[1;31mTimeoutError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import asyncio\n",
    "import json\n",
    "import websockets\n",
    "\n",
    "async def generate_text():\n",
    "    async with websockets.connect('wss://chat.petals.dev/api/v2/generate') as websocket:\n",
    "        # Send open_inference_session request\n",
    "        open_session_request = {\n",
    "            \"type\": \"open_inference_session\",\n",
    "            \"max_length\": 1024\n",
    "        }\n",
    "        await websocket.send(json.dumps(open_session_request))\n",
    "        \n",
    "        response = await websocket.recv()\n",
    "        response_data = json.loads(response)\n",
    "        \n",
    "        if response_data.get(\"ok\"):\n",
    "            print(\"Inference session opened successfully.\")\n",
    "        else:\n",
    "            print(\"Failed to open inference session:\", response_data.get(\"traceback\"))\n",
    "            return\n",
    "        \n",
    "        # Send generate request\n",
    "        generate_request = {\n",
    "            \"type\": \"generate\",\n",
    "            \"inputs\": \"A cat in French is \\\"\",\n",
    "            \"max_new_tokens\": 3\n",
    "        }\n",
    "        await websocket.send(json.dumps(generate_request))\n",
    "        \n",
    "        while True:\n",
    "            response = await websocket.recv()\n",
    "            response_data = json.loads(response)\n",
    "            \n",
    "            if response_data.get(\"ok\"):\n",
    "                outputs = response_data.get(\"outputs\")\n",
    "                stop = response_data.get(\"stop\")\n",
    "                print(\"Generated:\", outputs)\n",
    "                if stop:\n",
    "                    print(\"Generation completed.\")\n",
    "                    break\n",
    "            else:\n",
    "                print(\"Failed to generate text:\", response_data.get(\"traceback\"))\n",
    "                break\n",
    "\n",
    "# Run the async function using await\n",
    "await generate_text()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "llm_app_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
